### PR TITLE
File Downloader Submodule Major Refactor (Hi3Helper.Http)

### PR DESCRIPTION
**Pull request by neon-nyan**
# Main Goal
As per title, this PR contains some huge changes to our own File Downloader Submodule implementation that we use for Collapse called: Hi3Helper.Http.
The changes brings some new features to improve the reliability of the downloading process for certain parts, like Game Installation/Update, Game Repair, Cache Updates and Internal APIs. The changes that worth to mention are including:
- **[New]** Bandwidth/Speed Limiter
  With this new feature, the user can now set the download speed limit.
- **[New]** Download File Size Pre-allocation.
  Thanks to this feature, the download file would not need to have merged before the file can be used.
- **[Imp][Performance]** Reduce unnecessary ``async`` thread overhead.
- **[Imp]** Use delegates in replacement for event-based ``DownloadEvent`` to notify the download progress.
  This has benefits if you're performing download multiple files within a single Downloader instance for other purposes. By using delegates, you can subscribe the download progress to a specific file you wanted to track.
- **[Imp]** Changing the language specification from C# 8.0 to C# 9.0.

However, this PR requires some changes in some parts, including:
- [x] Game Repair
  - [x] Genshin Impact
  - [x] Honkai Impact 3rd
  - [x] Honkai: Star Rail
- [x] Cache Update
  - [x] Honkai Impact 3rd
  - [x] Honkai: Star Rail
- [x] Launcher API/Sprites Loader
  - [x] HoYoPlay
  - [x] Sophon
- [x] ``FallbackCDNUtil``
- [x] Game Conversion Utils
- [x] ``InstallManagerBase``
  - [x] Genshin Impact
  - [x] Honkai Impact 3rd
  - [x] Honkai: Star Rail
  - [x] Zenless Zone Zero
- [x] Launcher Updater

## PR Status :
- Overall Status : Completed
- Commits : Completed
- Synced to base (Collapse:main) : Not yet
- Build status : OK
- Crashing : No
- Bug found caused by PR : -

### Templates

<details>
  <summary>Changelog Prefixes</summary>
  
  ```
    **[New]**
    **[Imp]**
    **[Fix]**
    **[Loc]**
    **[Doc]**
  ```

</details>

### CODE REVIEW

File: CollapseLauncher/Classes/CachesManagement/Honkai/Fetch.cs [hunk 1]
```diff
@@ -34,57 +34,47 @@ private async Task<List<CacheAsset>> Fetch(CancellationToken token)
...
-                        case CacheAssetType.Unused:
-                        case CacheAssetType.Dispatcher:
-                        case CacheAssetType.Gateway:
-                        case CacheAssetType.General:
-                        case CacheAssetType.IFix:
-                        case CacheAssetType.DesignData:
-                        case CacheAssetType.Lua:
-                            continue;
-                    }
+                    case CacheAssetType.Unused:
```
Comment (bagusnl) [inline CollapseLauncher/Classes/CachesManagement/Honkai/Fetch.cs:49]: Do we need to download the unused cache? If yes then remove the comment at L#46

File: CollapseLauncher/Classes/Interfaces/Class/ProgressBase.cs [hunk 2]
```diff
@@ -407,6 +540,9 @@ protected async Task DoCopyStreamProgress(Stream source, Stream target, Cancella
 
         protected string EnsureCreationOfDirectory(string str)
         {
+            if (string.IsNullOrEmpty(str))
+                return str;
```
Comment (bagusnl) [inline CollapseLauncher/Classes/Interfaces/Class/ProgressBase.cs:544]: Potential null return.
Prolly just better to throw here

File: Hi3Helper.Core/Classes/Data/Tools/GenshinDispatchHelper/GenshinDispatchHelper.cs [hunk 3]
```diff
@@ -26,7 +26,7 @@ public class GenshinDispatchHelper : IDisposable
         public GenshinDispatchHelper(int RegionID, string DispatchKey, string DispatchURLPrefix, string VersionString = "2.6.0", CancellationToken cancelToken = new CancellationToken())
         {
             if (RegionID >= 4) ChannelName = "CNRELWin";
-            this._httpClient = new Http.Http(false, 1, 1);
+            this._httpClient = new Http.Legacy.Http(false, 1, 1);
```
Comment (bagusnl) [inline Hi3Helper.Core/Classes/Data/Tools/GenshinDispatchHelper/GenshinDispatchHelper.cs:29]: This is unimportant but if we dont really need this to be super fast or anything, I think we can move this to c#'s own httpclient, just less codes that needs to be monitored

File: Hi3Helper.Core/packages.lock.json [hunk 4]
```diff
@@ -4,8 +4,8 @@
     "net8.0": {
       "Google.Protobuf": {
         "type": "Transitive",
-        "resolved": "3.27.2",
-        "contentHash": "0wdgA3LO9mBS477jieBFs4pU1sWhVtwv/P+i9nAEiFDQyUA7PPHDBbJL1CeqYtV18jLiq9og4n7wSVCO171OBg=="
+        "resolved": "3.27.3",
```
Comment (bagusnl) [inline Hi3Helper.Core/packages.lock.json:7]: Its fine to up this to 3.28.0

**Review by bagusnl — CHANGES_REQUESTED** [review]
Mostly codeqa and stuff to talk about but overall its all ok

**Review by Cryotechnic — CHANGES_REQUESTED** [review]
Overall, PR looks OK and works as intended (tested locally). We could cleanup the magic numbers that we use in the code, such as for specifying max thread count and such by declaring them as variables in the function but functionally it works as intended (mostly for debugging later, to make it easier). That being said, the following UI issue is encountered for Cache Repair in Hi3 GLB region. I don't have the other regions installed so cannot verify if it is region agnostic:

image

Furthermore, please make sure to fix the i18n issues before merging. Aside from that, great work!

File: CollapseLauncher/Classes/CachesManagement/Honkai/Fetch.cs (hunk 1 above)
Comment (neon-nyan) [reply CollapseLauncher/Classes/CachesManagement/Honkai/Fetch.cs:49]: This is used to ignore the other types than Data, Event and AI (Asset Index).
Also, the reason why there are lots of Enum variant is that due to ``CacheAssetType`` being used for HSR and they have different types of data as well.

File: Hi3Helper.Core/packages.lock.json (hunk 4 above)
Comment (neon-nyan) [reply Hi3Helper.Core/packages.lock.json:7]: I will update the NuGet on Hi3Helper.Core once this PR is merged

**Comment by neon-nyan** [comment]
> Overall, PR looks OK and works as intended (tested locally). We could cleanup the magic numbers that we use in the code, such as for specifying max thread count and such by declaring them as variables in the function but functionally it works as intended (mostly for debugging later, to make it easier). That being said, the following UI issue is encountered for Cache Repair in Hi3 GLB region. I don't have the other regions installed so cannot verify if it is region agnostic:
> 
> image
> 
> Furthermore, please make sure to fix the i18n issues before merging. Aside from that, great work!

The issue with counts has a relation with the fetching mechanism to be enumerating the assets on-the-air rather than getting all the list of the assets and placed it into ``List<T>``. Since it's using ``IEnumerable<T>`` to enumerate the assets OTA, getting the total count is not possible because technically ``IEnumerable<T>`` doesn't have a ``Count`` property.

**Review by bagusnl — APPROVED** [review]
All gud i aprob

**Review by Cryotechnic — APPROVED** [review]
LGTM! Send it \o/

**Comment by zakidze** [comment]
can you show some pictures on how to set the limit , also is it released yet on stable version????

**Pull request merged**